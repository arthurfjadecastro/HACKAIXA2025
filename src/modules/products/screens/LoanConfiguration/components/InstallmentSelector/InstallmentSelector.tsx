import React from 'react';
import { View, Text, TouchableOpacity, TextInput } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { styles } from '../../LoanConfiguration.styles';

interface InstallmentSelectorProps {
  months: number;
  inputValue: string;
  productMinMonths: number;
  productMaxMonths: number;
  showValidationWarning: boolean;
  onDecrease: () => void;
  onIncrease: () => void;
  onChangeText: (text: string) => void;
  onBlur: () => void;
}

export const InstallmentSelector: React.FC<InstallmentSelectorProps> = ({
  months,
  inputValue,
  productMinMonths,
  productMaxMonths,
  showValidationWarning,
  onDecrease,
  onIncrease,
  onChangeText,
  onBlur,
}) => {
  return (
    <View style={styles.installmentSelector}>
      <Text style={styles.sectionLabel}>Parcelas</Text>
      
      <View style={styles.controlsRow}>
        <TouchableOpacity 
          onPress={onDecrease}
          style={[
            styles.controlButton, 
            months <= productMinMonths && styles.controlButtonDisabled
          ]}
          disabled={months <= productMinMonths}
        >
          <Ionicons 
            name="remove" 
            size={20} 
            color={months <= productMinMonths ? "#CCCCCC" : "#005CA9"} 
          />
        </TouchableOpacity>

        <TextInput
          style={styles.monthsInput}
          value={inputValue}
          onChangeText={onChangeText}
          onBlur={onBlur}
          keyboardType="number-pad"
          maxLength={3}
          textAlign="center"
        />

        <TouchableOpacity 
          onPress={onIncrease}
          style={[styles.controlButton, months >= productMaxMonths && styles.controlButtonDisabled]}
          disabled={months >= productMaxMonths}
        >
          <Ionicons name="add" size={20} color={months >= productMaxMonths ? "#CCCCCC" : "#005CA9"} />
        </TouchableOpacity>
      </View>

      <View style={styles.limitsHelper}>
        <Text style={{
          ...styles.helperText,
          color: showValidationWarning ? '#FF0000' : styles.helperText.color
        }}>
          Mín: {productMinMonths} • Máx: {productMaxMonths}
        </Text>
      </View>
    </View>
  );
};
