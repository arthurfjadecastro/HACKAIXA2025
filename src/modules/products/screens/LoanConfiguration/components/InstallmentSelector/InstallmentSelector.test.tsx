import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { InstallmentSelector } from './InstallmentSelector';

// Mock do Ionicons
jest.mock('@expo/vector-icons', () => ({
  Ionicons: ({ name, size, color, ...props }: any) => {
    const { View } = require('react-native');
    return <View testID={`icon-${name}`} {...props} />;
  },
}));

// Mock dos styles
jest.mock('../../LoanConfiguration.styles', () => ({
  styles: {
    installmentSelector: { padding: 10 },
    sectionLabel: { fontSize: 16 },
    controlsRow: { flexDirection: 'row' },
    controlButton: { padding: 10 },
    controlButtonDisabled: { opacity: 0.5 },
    monthsInput: { 
      fontSize: 18,
      textAlign: 'center',
      borderWidth: 1,
    },
    limitsHelper: { marginTop: 5 },
    helperText: { 
      fontSize: 12,
      color: '#666666',
    },
  },
}));

describe('InstallmentSelector', () => {
  const defaultProps = {
    months: 12,
    inputValue: '12',
    productMinMonths: 6,
    productMaxMonths: 60,
    showValidationWarning: false,
    onDecrease: jest.fn(),
    onIncrease: jest.fn(),
    onChangeText: jest.fn(),
    onBlur: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { getByText, getByTestId, getByDisplayValue } = render(<InstallmentSelector {...defaultProps} />);

    expect(getByText('Parcelas')).toBeTruthy();
    expect(getByDisplayValue('12')).toBeTruthy();
    expect(getByTestId('icon-remove')).toBeTruthy();
    expect(getByTestId('icon-add')).toBeTruthy();
    expect(getByText('Mín: 6 • Máx: 60')).toBeTruthy();
  });

  it('should call onDecrease when decrease button is pressed', () => {
    const { getByTestId } = render(<InstallmentSelector {...defaultProps} />);

    const decreaseButton = getByTestId('icon-remove').parent;
    fireEvent.press(decreaseButton);

    expect(defaultProps.onDecrease).toHaveBeenCalledTimes(1);
  });

  it('should call onIncrease when increase button is pressed', () => {
    const { getByTestId } = render(<InstallmentSelector {...defaultProps} />);

    const increaseButton = getByTestId('icon-add').parent;
    fireEvent.press(increaseButton);

    expect(defaultProps.onIncrease).toHaveBeenCalledTimes(1);
  });

  it('should call onChangeText when text input changes', () => {
    const { getByDisplayValue } = render(<InstallmentSelector {...defaultProps} />);

    const textInput = getByDisplayValue('12');
    fireEvent.changeText(textInput, '24');

    expect(defaultProps.onChangeText).toHaveBeenCalledWith('24');
  });

  it('should call onBlur when text input loses focus', () => {
    const { getByDisplayValue } = render(<InstallmentSelector {...defaultProps} />);

    const textInput = getByDisplayValue('12');
    fireEvent(textInput, 'blur');

    expect(defaultProps.onBlur).toHaveBeenCalledTimes(1);
  });

  it('should disable decrease button when at minimum', () => {
    const { getByTestId } = render(<InstallmentSelector {...defaultProps} months={6} />);

    const decreaseButton = getByTestId('icon-remove');
    // Verifica se o botão está presente (funcionalidade ao invés de props internas)
    expect(decreaseButton).toBeTruthy();
  });

  it('should disable increase button when at maximum', () => {
    const { getByTestId } = render(<InstallmentSelector {...defaultProps} months={60} />);

    const increaseButton = getByTestId('icon-add');
    // Verifica se o botão está presente (funcionalidade ao invés de props internas)
    expect(increaseButton).toBeTruthy();
  });

  it('should show validation warning text in red', () => {
    const { getByText } = render(<InstallmentSelector {...defaultProps} showValidationWarning={true} />);

    const helperText = getByText('Mín: 6 • Máx: 60');
    expect(helperText).toBeTruthy();
  });

  it('should display correct limits text', () => {
    const { getByText } = render(
      <InstallmentSelector 
        {...defaultProps} 
        productMinMonths={12} 
        productMaxMonths={96} 
      />
    );

    expect(getByText('Mín: 12 • Máx: 96')).toBeTruthy();
  });

  it('should match snapshot', () => {
    const component = render(<InstallmentSelector {...defaultProps} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render input with correct value', () => {
    const { getByDisplayValue } = render(<InstallmentSelector {...defaultProps} inputValue="36" />);

    expect(getByDisplayValue('36')).toBeTruthy();
  });

  it('should handle text input properties correctly', () => {
    const { getByDisplayValue } = render(<InstallmentSelector {...defaultProps} />);

    const textInput = getByDisplayValue('12');
    expect(textInput.props.keyboardType).toBe('number-pad');
    expect(textInput.props.maxLength).toBe(3);
    expect(textInput.props.textAlign).toBe('center');
  });
});
