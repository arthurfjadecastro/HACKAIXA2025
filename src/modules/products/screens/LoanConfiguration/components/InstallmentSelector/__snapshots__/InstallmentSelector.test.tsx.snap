// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstallmentSelector should match snapshot 1`] = `
<View
  style={
    {
      "padding": 10,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 16,
      }
    }
  >
    Parcelas
  </Text>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <View
        testID="icon-remove"
      />
    </View>
    <TextInput
      keyboardType="number-pad"
      maxLength={3}
      onBlur={[MockFunction]}
      onChangeText={[MockFunction]}
      style={
        {
          "borderWidth": 1,
          "fontSize": 18,
          "textAlign": "center",
        }
      }
      textAlign="center"
      value="12"
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <View
        testID="icon-add"
      />
    </View>
  </View>
  <View
    style={
      {
        "marginTop": 5,
      }
    }
  >
    <Text
      style={
        {
          "color": "#666666",
          "fontSize": 12,
        }
      }
    >
      Mín: 
      6
       • Máx: 
      60
    </Text>
  </View>
</View>
`;
